Consumption specifications for cost insights and reporting operations

shalinved opened the pull request:
This checklist is used to make sure that common issues in a pull request are addressed. This will expedite the process of getting your pull request merged and avoid extra work on your part to fix issues discovered during the review process.

### PR information
- [ ] The title of the PR is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For information on cleaning up the commits in your pull request, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).
- [ ] Except for special cases involving multiple contributors, the PR is started from a fork of the main repository, not a branch.
- [ ] If applicable, the PR references the bug/issue that it fixes.
- [ ] Swagger files are correctly named (e.g. the `api-version` in the path should match the `api-version` in the spec).

### Quality of Swagger
- [ ] I have read the [contribution guidelines](https://github.com/Azure/azure-rest-api-specs/blob/master/.github/CONTRIBUTING.md).
- [ ] My spec meets the review criteria:
  - [ ] The spec conforms to the [Swagger 2.0 specification](https://github.com/OAI/OpenAPI-Specification/blob/master/versions/2.0.md).
  - [ ] The spec follows the guidelines described in the [Swagger checklist](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md) document.
  - [ ] [Validation tools](https://github.com/Azure/azure-rest-api-specs/blob/master/documentation/swagger-checklist.md#validation-tools-for-swagger-checklist) were run on swagger spec(s) and have all been fixed in this PR.

Commit:
Consumption specifications for cost insights and reporting operations

msftclas commented:
[![CLA assistant check](https://cla.opensource.microsoft.com/pull/badge/not_signed)](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3171) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3171) before we can accept your contribution.<br/><br/>:x: shalinved [sign now](https://cla.opensource.microsoft.com/Azure/azure-rest-api-specs?pullRequest=3171)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.opensource.microsoft.com/check/Azure/azure-rest-api-specs?pullRequest=3171) it.</sub>

AutorestCI commented:
# Automation for azure-libraries-for-java

AutorestCI commented:
# Automation for azure-sdk-for-python

AutorestCI commented:
# Automation for azure-sdk-for-node

AutorestCI commented:
# Automation for azure-sdk-for-ruby

AutorestCI commented:
# Automation for azure-sdk-for-go

lmazuel commented:
Hi @shalinved 
Since it's a new API, adding @ravbhatnagar for ARM feedback
Please also fix all CI issues. For instance:
> ERROR: Referenced file 'file:///tmp/tmp6a6_3m9v/rest/specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-05-31/e./examples/ReportConfigDelete.json' not found

Thanks!

Commit:
Fixed $ref for ReportConfigDelete

shalinved commented:
@lmazuel - I just pushed a commit to fix bad reference for ReportConfigDelete.
@ravbhatnagar - Hi, Gaurav, these APIs are the ones we talked about some time back and had a casual discussion about. Let me know if you would like to schedule a formal review meeting.

Commit:
Silly me!! Fixed bad reference again

Commit:
Fixed schema validation issues

Commit:
Fixing validation errors around datetime and dynamic respones

Commit:
Fixed parameters in examples

Commit:
Fixed parameter name in example

lmazuel commented:
@fearthecowboy Autorest fails saying this:
> FATAL: System.InvalidOperationException: Method with x-ms-odata needs to have "$filter" parameter.

But all linter rules are fine and green. Is this a linter problem? Autorest problem?

fearthecowboy commented:
I would think there needs to be a linter rule for this; https://github.com/Azure/autorest.common/blob/e85fe8a6a6c89a62c93708f7da4a8c555590c436/src/AzureExtensions.cs shows that it expects the `$filter` ...

Commit:
Removed x-ms-odata method for POST op

ravbhatnagar commented on specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-05-31/consumption.json:
```diff
@@ -120,115 +120,1382 @@
+        }
+      }
+    },
+    "/subscriptions/{subscriptionId}/providers/Microsoft.Consumption/reportconfigs/{reportConfigName}": {
```
Is reportConfigs created under a resource group or under a subscription?

ravbhatnagar replied:
Shouldnt it be created under an RG

shalinved replied:
We allow report configurations to be created under a subscription as well as under a RG. This determines scope of the data in the report. This is similar to what we do currently in Consumption Budgets as well.

ravbhatnagar commented on specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-05-31/consumption.json:
```diff
@@ -120,115 +120,1382 @@
+          "type": "string",
+          "description": "Resource type."
+        },
+        "tags": {
```
why this is readonly?

ravbhatnagar reviewed (commented):
@shalinved  - lets schedule a quick sync to go over these. I have a few questions about these. Also, please loop in folks from azureapirbcore - @johanste @devigned @fearthecowboy

ravbhatnagar commented:
Had a call with Shalin - 3 things -
1. POST APIs for insights - change the name to represent an action - ex- generateInsight. We discussed about making these PUT and enabling GET. Currently someone needs this action permission to call this API. Reader cant get the insight. RP is fine with this behavior
2. Reference to storage account should be through resourceID with Linked access check enabled. - blocking feedback
3. Make sure the size of data coming back in response for dimensions is under the ARM limits. As it can grow, it can cause problems.

ms-premp commented:
adding ms-premp

ms-premp commented:
adding @ms-premp

ravbhatnagar commented:
@shalinved - any updates on this? @ms-premp ?

shalinved commented:
@ravbhatnagar - Resource id changes are done. I am waiting for a decision on the name of Insights API.

shalinved commented:
@marbing - Adding Matt to the discussion

Commit:
Fixed ARM review comments

Commit:
Merging upstream to fork

shalinved commented:
@ravbhatnagar - I have pushed changes that would take care of points 1 and 2 above.
I have modified our request contract to accept an ARM resource id for the storage account where reports will be delivered. We will use ARM linked access check to validate user has read permissions on the resource while creating a report configuration.
I have also updated the Insights API to `AnalyzeUsage`.
We do have checks in place to make sure we don't go above ARM limits for Dimensions API. The API is pageable with a page size of 1000 records.
Could you please review the changes.

ravbhatnagar commented:
looks good

lmazuel commented on specification/consumption/resource-manager/Microsoft.Consumption/stable/2018-05-31/consumption.json:
```diff
@@ -1324,148 +2066,653 @@
+          "format": "date-time",
+          "type": "string"
+        },
+        "nextLink": {
```
Why would this object get a nextLink? Just checking that it's accurate, since it's not usual and SDK want be able to do anything with that, and polling would be manual.

ms-premp replied:
Hi @imazuel 
This is intentional. The response here is a single object and not a list of object. The single object however, can have a very big payload in data field. Hence, the need for pagination. We are aware that SDKs will not be able to parse it automatically and we will have additional work to support this in SDK. 
This pattern is already used in one other API v.i.z. Pricesheets. We had multiple back and forth discussions with ARM team and this is the only way we can come up with since the response here is a single ARM resource object and not an array of ARM resources.

lmazuel reviewed (commented):
One question to confirm something unusual, otherwise LGTM

jhendrixMSFT commented:
@AutorestCI regenerate azure-sdk-for-go
